### DISCUSSION TITLE
Welcome to SwissQRBill.NET Discussions!

### DISCUSSION THREAD

**Discussion by manuelbl** [manuelbl/SwissQRBill.NET#48] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of what you are doing (or trying to do) with QR bills.

**Comment by TimAllemann** [comment]
Hi, it's a really good library. Do you have any idea if a parser from a string exist ?

for example you read with you webcam and receive _SPC 0200 1 CH5800791123000889012 S Robert Schneider AG Rue du Lac 1268 2501 Bienne CH 2500.25 CHF S Pia-Maria Rutschmann-Schnyder Grosse Marktgasse 28 9400 Rorschach CH SCOR RF18539007547034 EPD Nom AV1: UV;UltraPay005;12345 Nom AV2: XY;XYService;54321_ 

then this would create you object in .net.

Or is this someting I should implement myself based on https://www.paymentstandards.ch/dam/downloads/ig-qr-bill-en.pdf

**Reply by LuFet** [reply]
Hallo Manuel
was kann ich machen, wenn ich eine ältere version von Net auf dem Server habe? Danke im Voraus!

**Reply by manuelbl** [reply]
Verstehe ich das richtig: Du möchtest SwissQRBill.NET einsetzen, aber der Server, auf dem die Anwendung laufen soll, läuft mit einer zu alten .NET-Version, also älter als .NET Framework 4.6.1?

Falls ja, sehe ich folgende Optionen:

- Server upgraden. Das wolltet ihr sicher schon lange machen, habt es aber immer wieder rausgeschoben.
- Den Source Code von SwissQRBill.NET unter der alten .NET-Version kompilieren. Vielleicht läuft es ja mit minimalen Anpassungen.
- Eine andere Software für die Erstellung von QR-Rechnungen kaufen oder entwickeln lassen.

**Reply by LuFet** [reply]
Vielen Dank!

**Comment by manuelbl** [comment]
Yes, that exists. Have a look at [QRBill.DecodeQrCodeText(String)](https://codecrete.net/SwissQRBill.NET/api/Codecrete.SwissQRBill.Generator.QRBill.html#Codecrete_SwissQRBill_Generator_QRBill_DecodeQrCodeText_System_String_)

**Reply by TimAllemann** [reply]
Sorry I didn't read the docs correctly... My bad. I will try that next week, thank you very much.

**Comment by maerger** [comment]
Hi there,

just tried to create a png bill using Codecrete.SwissQRBill.Windows (.net framework 4.6.2) and get the following Error (svg is running fine):

```
Test method Crm.QrBillGenerator.Tests.QrBillHelperTests.GetBillTest threw exception: 
Codecrete.SwissQRBill.Generator.QRBillGenerationException: Graphics format PNG not available (are you missing the NuGet package Codecrete.SwissQRBill.Generator?)

  Stack Trace: 
QRBill.CreateCanvas(BillFormat format) line 311
QRBill.Generate(Bill bill) line 116
```
As I read in the docs there is no need to get the Generator using the Windows version. Is this correct?

btw: The following nuget packages installed:
- Codecrete.SwissQRBill.Core version="3.0.4" targetFramework="net462"
- Codecrete.SwissQRBill.Windows version="3.0.4" targetFramework="net462"
- System.Drawing.Common version="6.0.0" targetFramework="net462"

**Reply by manuelbl** [reply]
The setup looks correct. I'll have to investigate it in more detail.

**Reply by maerger** [reply]
Tried with the following code:
```
            byte[] png;
            using (PNGCanvas canvas = new PNGCanvas(QRBill.QrBillWidth, QRBill.QrBillHeight, 300, "\"Liberation Sans\",Arial, Helvetica"))
            {
                bill.Format.OutputSize = OutputSize.QrBillOnly;
                QRBill.Draw(bill, canvas);
                png = canvas.ToByteArray();
            }

            return png;
```
Now i get the following Error:

> Message: 
> Test method Crm.QrBillGenerator.Tests.QrBillHelperTests.GetBillTest threw exception: 
> System.IO.FileNotFoundException: Could not load file or assembly 'System.Drawing.Common, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified. ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Drawing.Common, Version=4.0.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
> 
>   Stack Trace: 
> BitmapCanvas.ctor(Int32 width, Int32 height, Single dpi, String fontFamilyList)
> PNGCanvas.ctor(Double width, Double height, Single dpi, String fontFamilyList) line 29

ms there is an explicit ref to "System.Drawing.Common, Version=4.0.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51".

**Reply by manuelbl** [reply]
I'm afraid I can't reproduce the error. I tried different combinations of NuGet references and .NET framework version, and they all work.

In your case, a reference to *Codecrete.SwissQRBill.Windows* should be sufficient. To other references are resolved indirectly.

The error message re *"Could not load ... System.Drawing.Common ..."* is very strange. It could indicate a broken project or .NET installation. And version 4.0.0.1 is very low. To the best of my knowledge, it predates .NET Framework 4.6.2.

Can you try to create a new console app with .NET Framework, add the NuGet package *Codecrete.SwissQRBill.Windows* and then replace the *Program.cs* with this code: https://github.com/manuelbl/SwissQRBill.NET/blob/master/Examples/Basic/Program.cs. Does it run?

**Reply by maerger** [reply]
Thanks for the answer - yes it does. When I put the same code into my other project and get the following error:
Message: 

> Test method Crm.QrBillGenerator.Tests.QrBillHelperTests.GetBillTest threw exception: 
> Codecrete.SwissQRBill.Generator.QRBillGenerationException: Graphics format PNG not available (are you missing the NuGet package Codecrete.SwissQRBill.Generator?)
> 
>   Stack Trace: 
> QRBill.CreateCanvas(BillFormat format) line 311
> QRBill.Generate(Bill bill) line 116
> QrBillHelper.GeneratePngBill(Bill bill) line 72

Seems there is something wrong with my existing project... I will try to figure out.

**Reply by manuelbl** [reply]
This error message likely covers the underlying cause, which is that System.Drawing.Common cannot be loaded.

**Reply by maerger** [reply]
Changed my assembly project to executable and copied your code - it is running fine. If I call the same method from my test project, I get the error.

**Reply by manuelbl** [reply]
I can reproduce it with a .NET framework unit test project. Other than for newer types of test project, this type does not reference all required dependencies. In particular, it does not reference `Codecrete.SwissQRBill.Windows`. This NuGet package is special as it extends the *Core* package and is loaded dynamically. This seems to make a difference for the .NET framework unit test project. If you add it explicitly to the test project, it works.

**Reply by maerger** [reply]
Found a cleaner solution adding the line:

` CanvasCreator.Register((ICanvasFactory)new PNGCanvasFactory());`

Maybe considering a proper dependency injection... if I have the time, I will try to refactor the "RegisterPixelCanvasFactory()" Method in CanvasCreator.cs

**Reply by manuelbl** [reply]
For a test project, that's a good solution. But as a convenience for developers, `RegisterPixelCanvasFactory()` tries to make it work without requiring developers to add code (as they usually don't read the documentation). This wouldn't be needed if .NET had a standard cross platform library for pixel graphics. But with .NET 6, that's no longer the case.

Regarding dependency injection: I'm not aware of a cross DLL dependency injection solution available in .NET Standard 2.0. But if you are, I'm highly interested to hear about it.
